### PR TITLE
[4.0] Add the Report-To header to the http header config and allow to set the script-dynamic header

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Add the [Report-To](https://w3c.github.io/reporting/) ([#2](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-to)) header to the http header config and allow to set the [script-dynamic](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src) header 

### Testing Instructions

- go to the plg_system_httpheaders
- notice there is no report-to header in the `force header` option.
- apply this patch
- the option is there
- set the option to any value. ([or see here the details](https://w3c.github.io/reporting/) + [example](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-to))
- revert the patch
- go to the com_csp options
- notice there is no strict-dynamic setting
- apply this patch
- notice there is now a strict dynamatic setting. [(details)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src)
- enable the strict dynamic setting
- (when in custom mode make sure you setup a inital script-src dummy rule)
- and check the generated csp header. (F12 -> Network tab)

### Expected result

- you can set the report-to via the plugin
- you can set the strict-dynamic option for the CSP

### Actual result

- you can not set the report-to via the plugin
- you can not set the strict-dynamic option for the CSP

### Documentation Changes Required

The script-dynamic options needs to be added to doc pages the Report-To has been added already
https://docs.joomla.org/index.php?title=Help4.x:Components_CSP_Reports_Options
https://docs.joomla.org/J4.x:Http_Header_Management

### CODE REVIEW

File: plugins/system/httpheaders/httpheaders.php [hunk 1]
```diff
@@ -320,6 +322,14 @@ private function setCspHeader(): void
 					$frameAncestorsSet = true;
 				}
 
+				// Add strict-dynamic to the script-src directive when enabeld
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.php]:
```suggestion
				// Add strict-dynamic to the script-src directive when enabled
```

File: plugins/system/httpheaders/httpheaders.php [hunk 2]
```diff
@@ -442,6 +453,14 @@ private function compileAutomaticCspHeaderRules(): array
 				$cspHeaderValue = '{style-hashes} ' . $cspHeaderValue;
 			}
 
+			// Add strict-dynamic to the script-src directive when enabeld
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.php]:
```suggestion
			// Add strict-dynamic to the script-src directive when enabled
```

**Comment by zero-24** [comment]
Thanks @Quy

File: plugins/system/httpheaders/httpheaders.xml [hunk 3]
```diff
@@ -66,6 +66,7 @@
 							<option value="referrer-policy">Referrer-Policy</option>
 							<option value="expect-ct">Expect-CT</option>
 							<option value="feature-policy">Feature-Policy</option>
+							<option value="report-to">Report-to</option>
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.xml]:
```suggestion
							<option value="report-to">Report-To</option>
```

File: plugins/system/httpheaders/httpheaders.php [hunk 4]
```diff
@@ -320,6 +322,14 @@ private function setCspHeader(): void
 					$frameAncestorsSet = true;
 				}
 
+				// Add strict-dynamic to the script-src directive when enabled
+				if ($cspValue->directive === 'script-src'
+					&& strpos($cspValue->value, 'strict-dynamic') === 0
+					&& $scriptDynamicEnabled)
+				{
+					$cspValue->value = $cspValue->value . ' strict-dynamic';
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.php]:
```suggestion
					$cspValue->value .= ' strict-dynamic';
```

File: plugins/system/httpheaders/httpheaders.php [hunk 5]
```diff
@@ -442,6 +453,14 @@ private function compileAutomaticCspHeaderRules(): array
 				$cspHeaderValue = '{style-hashes} ' . $cspHeaderValue;
 			}
 
+			// Add strict-dynamic to the script-src directive when enabled
+			if ($cspHeaderkey === 'script-src'
+				&& strpos($cspHeaderValue, 'strict-dynamic') === 0
+				&& $scriptDynamicEnabled)
+			{
+				$cspHeaderValue = $cspHeaderValue . ' strict-dynamic';
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.php]:
```suggestion
				$cspHeaderValue .= ' strict-dynamic';
```

**Comment by zero-24** [comment]
Thanks merged @Quy

File: administrator/language/en-GB/com_csp.ini [hunk 6]
```diff
@@ -22,6 +22,8 @@ COM_CSP_CONTENTSECURITYPOLICY_NONCE_ENABLED_DESC="Enable the whitelist for speci
 COM_CSP_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="Use the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy'."
 ; Please do not translate the following language string
 COM_CSP_CONTENTSECURITYPOLICY_REPORT_ONLY="Report-Only"
+COM_CSP_CONTENTSECURITYPOLICY_SCRIPT_DYNAMIC_ENABLED="strict-dynamic"
```
Comment (brianteeman) [inline administrator/language/en-GB/com_csp.ini]: you need to add a note about not translating "strict-dynamic" in this string and the following one

Comment (zero-24) [reply administrator/language/en-GB/com_csp.ini]: thanks just done that.

File: administrator/language/en-GB/com_csp.ini [hunk 7]
```diff
@@ -20,8 +20,11 @@ COM_CSP_CONTENTSECURITYPOLICY_NONCE_ENABLED="<a href='https://scotthelme.co.uk/c
 COM_CSP_CONTENTSECURITYPOLICY_NONCE_ENABLED_DESC="Enable the whitelist for specific inline scripts using a cryptographic nonce (number used once) for all scripts and styles using the Joomla API. Specifying a nonce makes a modern browser ignore 'unsafe-inline' which should still be set for older browsers without nonce support."
 ; Please do not translate 'Content-Security-Policy' & 'Content-Security-Policy-Report-Only' in the following language string
 COM_CSP_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="Use the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy'."
-; Please do not translate the following language string
+; Please do not translate the following 2 language strings
 COM_CSP_CONTENTSECURITYPOLICY_REPORT_ONLY="Report-Only"
+COM_CSP_CONTENTSECURITYPOLICY_SCRIPT_DYNAMIC_ENABLED="strict-dynamic"
+; Please do not translate 'strict-dynamic' in the following language string
```
Comment (Quy) [inline administrator/language/en-GB/com_csp.ini]: Should `self` and `unsafe-inline` be included to not translate?

Comment (zero-24) [reply administrator/language/en-GB/com_csp.ini]: you are right missed them patched. 👍

**Comment by Quy** [comment]
Please fix conflicts.

**Comment by zero-24** [comment]
Done @Quy

File: plugins/system/httpheaders/httpheaders.php [hunk 8]
```diff
@@ -79,6 +79,7 @@ class PlgSystemHttpHeaders extends CMSPlugin implements SubscriberInterface
 		'expect-ct',
 		'feature-policy',
 		'cross-origin-opener-policy',
+    'report-to',
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.php]:
```suggestion
		'report-to',
```

**Comment by zero-24** [comment]
Any blockers other than testing at this point?

**Comment by richard67** [comment]
@zero-24 I see no blocker and just wanted to test, but I've noticed that Drone status is not ok here, it is hanging on "Waiting for status to be reported". I have no idea why, the branch seems not to have any conflicts to the 4.0-dev branch of the CMS.

**Comment by richard67** [comment]
@zero-24 Maybe you can fix the hanging drone by updating your branch to latest 4.0-dev?

**Comment by zero-24** [comment]
Done @richard67

**Comment by jmeintrup** [comment]
I have tested this item :white_check_mark: successfully on 72ba37cc01de3b8fa6aad2b3602f3493dafe5b74<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28724">issues.joomla.org/tracker/joomla-cms/28724</a>.</sub>

**Comment by zero-24** [comment]
Drone is ok nowdays so nothing bocks testing here right @richard67 ? Maybe @Quy can mark a test as the last commits where just branch updates the test by @jmeintrup still stands.

File: plugins/system/httpheaders/httpheaders.php [hunk 9]
```diff
@@ -325,6 +327,14 @@ private function setCspHeader(): void
 					$frameAncestorsSet = true;
 				}
 
+				// Add strict-dynamic to the script-src directive when enabled
+				if ($cspValue->directive === 'script-src'
+					&& strpos($cspValue->value, 'strict-dynamic') === 0
+					&& $scriptDynamicEnabled)
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.php]: Minor suggestion to check `$scriptDynamicEnabled` first.

Comment (zero-24) [reply plugins/system/httpheaders/httpheaders.php]: Done. https://github.com/joomla/joomla-cms/pull/28724/commits/64897e8a9dc53d34cdd6758c9f09fed641cfa5da

**Comment by Quy** [comment]
Is `script-dynamic` dependent on `Nonce` and/or `Script hashes` being enabled?

**Comment by zero-24** [comment]
> Is script-dynamic dependent on Nonce and/or Script hashes being enabled?

I dont think that it is a hard requiremet but suggested for sure.

**Comment by Quy** [comment]
How about make it like `frame-ancestors` so you don't have to add an entry for it?

**Comment by zero-24** [comment]
> How about make it like frame-ancestors so you don't have to add an entry for it?

I think it should be an informed decision to enable script-dynamic or not.

**Comment by Quy** [comment]
My suggestion is to keep the `script-dynamic` switch. If enabled, `strict-dynamic` will be added.

**Comment by zero-24** [comment]
Agree that should be the case right now? Or I'm missing something?

**Comment by Quy** [comment]
I assume you have to add a Directive entry to activate it.

**Comment by zero-24** [comment]
> I assume you have to add a Directive entry to activate it.

Ah i get where you are comming from. `frame-ancestors` can be compared to `script-src` in that case it is selfcontained. `scrypt-dynamic` should be passed as part of the `script-src` directive.

**Comment by Quy** [comment]
I am new to CSP so I don't understand it fully, however, I think Yes to your assessment.

**Comment by zero-24** [comment]
> I am new to CSP so I don't understand it fully, however, I think Yes to your assessment.

no problem feel free to ask when questions like that come up :)

File: administrator/components/com_csp/config.xml [hunk 10]
```diff
@@ -71,6 +71,18 @@
 			<option value="0">JDISABLED</option>
 			<option value="1">JENABLED</option>
 		</field>
+		<field
+			name="strict_dynamic_enabled"
+			type="radio"
+			label="COM_CSP_CONTENTSECURITYPOLICY_SCRIPT_DYNAMIC_ENABLED"
```
Comment (wilsonge) [inline administrator/components/com_csp/config.xml]: Can we standardize these based on the field name - either script dynamic or strict dynamic but not some sort of hybrid?

**Comment by opn365** [comment]
I have tested this item :red_circle: unsuccessfully on 64897e8a9dc53d34cdd6758c9f09fed641cfa5da<br /><br />the test says : notice there is no script-dynamic setting, the option in j4 says strict-dynamic   
so either one or the other is inconsistent  (script- or strict- ... which one? )<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28724">issues.joomla.org/tracker/joomla-cms/28724</a>.</sub>

**Comment by zero-24** [comment]
@opn365 this has been patched in the description now. It was an typo in the description please retest.

File: administrator/components/com_csp/config.xml (hunk 10 above)
Comment (zero-24) [reply administrator/components/com_csp/config.xml]: Done.

**Comment by opn365** [comment]
I have tested this item :white_check_mark: successfully on 02826b83a89fee55c470f2e3144ae515dc676542<br /><br />Test successful after description update<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28724">issues.joomla.org/tracker/joomla-cms/28724</a>.</sub>

**Comment by Bodge-IT** [comment]
@zero-24 during test the new option in the com_csp is not appearing.  something i'm missing?

**Comment by zero-24** [comment]
> @zero-24 during test the new option in the com_csp is not appearing. something i'm missing?

hmm can you show me the options screen? And please double check that the correct PR was applied.

**Comment by Bodge-IT** [comment]
image
@zero-24 pr correct as first section of test works

**Comment by zero-24** [comment]
> @zero-24 pr correct as first section of test works

please enable that option and configure "auto" or "custom" mode to see the new option.

**Comment by Bodge-IT** [comment]
> 
> 
> > @zero-24 pr correct as first section of test works
> 
> please enable that option and configure "auto" or "custom" mode to see the new option.

OK, got to the setting, activated it and tested on https://csp-evaluator.withgoogle.com/%7C
getting:
image

**Comment by zero-24** [comment]
> OK, got to the setting, activated it and tested on https://csp-evaluator.withgoogle.com/%7C
getting:

Have you setup the script-src and object-src? What are the settings from the component?

**Comment by Bodge-IT** [comment]
OK, better question how do I check for "the generated csp header."

**Comment by richard67** [comment]
@Bodge-IT In development tools of your browser, network analysis, then inspect the header of the packet. It's something for nerds ;-)

**Comment by Bodge-IT** [comment]
thanks @richard67 , 

referrer-policy | strict-origin-when-cross-origin

This what I'm looking for? Am new to CSP.

**Comment by richard67** [comment]
> referrer-policy | strict-origin-when-cross-origin
> 
> This what I'm looking for? Am new to CSP.

At this point @zero-24 can help netter than I do.

**Comment by zero-24** [comment]
> This what I'm looking for? Am new to CSP.

Conent-security-poliy(-report-only)

Can you please post the settings you made?

**Comment by Bodge-IT** [comment]
com_csp:
image

**Comment by Bodge-IT** [comment]
Plugin:
image

**Comment by Bodge-IT** [comment]
headers:
image

**Comment by richard67** [comment]
@Bodge-IT In your screenshot of the headers I see: "report-to: csp-endpoint". That's one of the headers added by this PR, so success for this point.

For the second point the strict-dynamic option option, your screenshot with the com_csp options shows that this option is there. So success also for this point.

@zero-24 Am I right? Or am I missing something?

**Comment by zero-24** [comment]
> @zero-24 Am I right? Or am I missing something?

Yes @Bodge-IT in custom mode you can actually also add a custom rule ("last option") with script-src and some value and notice that the CSP header has that key value as well as script-dynamic.

**Comment by Bodge-IT** [comment]
I get this:
image

**Comment by zero-24** [comment]
Ok found the issue please re apply this PR (revert and than apply again via patchtester) or manually apply this changes here: https://github.com/joomla/joomla-cms/pull/28724/commits/304c46808a7bb16509d877dfe724243a66dd3e10

**Comment by Bodge-IT** [comment]
Boom:
image

 Will reset @opn365 test and request new additional test

Thanks for your support...

**Comment by zero-24** [comment]
Thanks 👍

**Comment by Bodge-IT** [comment]
Do we need to get tests reset?

@zero-24, can you tweak test instructions to advise what we're looking for in headers?

**Comment by richard67** [comment]
@Bodge-IT No need for reset, that happens automatically on GitHub with a new commit. In the tracker you might not see that. But it's here on GitHub which counts.

@opn365 Could you re apply this PR (revert, then fetch again patches and than apply again via patchtester if using patchtester, or pulling latest changes when using a git client) and then repeat your test? There have been changes made in this PR. Thanks in advance, and thanks for the previous test.

**Comment by zero-24** [comment]
> @zero-24, can you tweak test instructions to advise what we're looking for in headers?

Done thanks

**Comment by Bodge-IT** [comment]
I have tested this item :white_check_mark: successfully on 80d1e1bf1131926f1907b5c9f3ecf1b2e9750954<br /><br />Checked options and headers...all good.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28724">issues.joomla.org/tracker/joomla-cms/28724</a>.</sub>

**Comment by flo-the-cat** [comment]
I have tested this item :white_check_mark: successfully on 80d1e1bf1131926f1907b5c9f3ecf1b2e9750954<br /><br />Was easy as pie to follow, once Gary added all the screen shots!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28724">issues.joomla.org/tracker/joomla-cms/28724</a>.</sub>

**Comment by Bodge-IT** [comment]
Pulled in Phil to get this done...

**Comment by richard67** [comment]
> Pulled in Phil to get this done...

Seems he did not have time and has sent his cat instead ;-)

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28724">issues.joomla.org/tracker/joomla-cms/28724</a>.</sub>

**Pull request merged**